### DISCUSSION TITLE
State of FormControl

### DISCUSSION THREAD

**Discussion by Jonas-C** [chakra-ui/ark#2525] [General]
Hi! According to your [roadmap](https://ark-ui.canny.io/), a Form/FormControl component has been planned for a while now. I saw that the chakra-ui migration to ark is still using its own implementation of FormControl, and was wondering about the progress of a potential Ark FormControl. Is it in the works? Would it be released with the next major version?

Most, if not all of your components already support automatically grouping labels and their respective components. Furthermore, it seems like several of your components support `required`, `invalid` and `readonly`. Are there any plans to export `ErrorMessage` and `HelperText`components for the form components?

**Comment by segunadebayo** [comment]
@Jonas-C Yes, we have plans to add support for those Form-related components. Most of the code in Chakra will end up in Ark.

> Side note: It's really exciting to see what you're building in NDLA frontend packages repo 👌🏼

**Reply by Jonas-C** [reply]
Thanks, and thanks! I'm excited to use Ark! 

Do you have any indication as to what it'll look like? Right now we're currently using something that looks a whole lot like chakras FormControl. I'm considering creating small wrappers around the form-related components in Ark now where I'll consume the FormControl context and pass it down to the ark root components. Does that seem like an approach that could easily be switched over to the ark FormControl in the future?

**Reply by cschroeter** [reply]
@Jonas-C 

Here is what we have on the whiteboard. Happy to hear your feedback.


### Interface

```ts
interface FieldProps {
  id?: string
  invalid?: boolean;
  disabled?: boolean;
  readOnly?: boolean;
}
```

### Using Ark Components

```jsx
<Field.Root>
    <Select.Root>
        <Select.Label>Label</Select.Label>
    </Select.Root>
    <Field.HelperText>Some info</Field.HelperText>
    <Field.ErrorText>Some info</Field.ErrorText>
</Field.Root>
```

### Custom Components
```jsx
<Field.Root>
    <Field.Label>Email address</Field.Label>
    // <Field.Context>
    //     {api=> <input type="email" {...api.getControlProps()} />}
    // </Field.Context>
    <Field.Control asChild>
        <input type="email" />
    </Field.Control>
    <Field.HelperText>Some info</Field.HelperText>
    <Field.ErrorText>Some info</Field.ErrorText>
</Field.Root>

**Reply by Jonas-C** [reply]
This looks great, thank you so much! I'll keep making shallow wrappers until this lands 😄

**Reply by afkcodes** [reply]
So Ark's Field and FormControl from Chakra is pretty much the same is it ? @segunadebayo @cschroeter
